csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll

```diff
@@ -496,98 +497,137 @@ module ControlFlow {
+        not e instanceof QualifiableExpr and
+        not e instanceof Assignment and
+        not e instanceof AnonymousFunctionExpr and
+        result = e.getChild(i)
         or
-        i = 1 and result = ie.(IsPatternExpr).getVariableDeclExpr()
+        e = any(ExtensionMethodCall emc | result = emc.getArgument(i))
         or
-        i = 1 and result = ie.(IsConstantExpr).getConstant()
+        e = any(QualifiableExpr qe |
+            not qe instanceof ExtensionMethodCall and
+            not qe.isConditional() and
+            result = qe.getChild(i)
+          )
+        or
+        e = any(Assignment a |
+            // The left-hand side of an assignment is evaluated before the right-hand side
+            i = 0 and result = a.getLValue()
+            or
+            i = 1 and result = a.getRValue()
+          )
       }
 
-      private ControlFlowElement getAsExprChild(AsExpr ae, int i) {
-        // The type access at index 1 is not evaluated at run-time
-        i = 0 and result = ae.getExpr()
+      private ControlFlowElement getExprChildElement(Expr e, int i) {
+        result = rank[i + 1](ControlFlowElement cfe, int j |
+            cfe = getExprChildElement0(e, j) and
+            not cfe instanceof NoNodeExpr
+          |
+            cfe order by j
+          )
       }
 
-      private ControlFlowElement getUnknownExprChild(@unknown_expr e, int i) {
-        exists(int c | result = e.(Expr).getChild(c) |
-          c = rank[i + 1](int j | exists(e.(Expr).getChild(j)))
-        )
+      private int getFirstChildElement(Expr e) {
+        result = min(int i | exists(getExprChildElement(e, i)))
       }
 
-      private ControlFlowElement getCastExprChild(CastExpr ce, int i) {
-        // The type access at index 1 is not evaluated at run-time
-        i = 0 and result = ce.getExpr()
+      private int getLastChildElement(Expr e) {
+        result = max(int i | exists(getExprChildElement(e, i)))
+      }
+
+      /**
+       * A qualified write access. In a qualified write access, the access itself is
+       * not evaluated, only the qualifier and the indexer arguments (if any).
+       */
+      private class QualifiedWriteAccess extends WriteAccess, QualifiableExpr {
+        QualifiedWriteAccess() { this.hasQualifier() }
+      }
+
+      /**
+       * An expression that writes via an accessor call, for example `x.Prop = 0`,
+       * where `Prop` is a property.
+       *
+       * Accessor writes need special attention, because we need to model the fact
+       * that the accessor is called *after* the assigned value has been evaluated.
+       * In the example above, this means we want a CFG that looks like
+       *
+       * ```
+       * x -> 0 -> set_Prop -> x.Prop = 0
+       * ```
+       */
+      class AccessorWrite extends Expr {
+        AssignableDefinition def;
+
+        AccessorWrite() {
+          def.getExpr() = this and
+          def.getTargetAccess().(WriteAccess) instanceof AccessorCall and
```

Comment:
Not sure the trailing `(WriteAccess)` is needed

Reply:
It is required to avoid writes such as `x.Prop++`, where the access `x.Prop` is both a read and a write.